### PR TITLE
Enhanced extraction API

### PR DESCRIPTION
**Pull request by tombujok**
Enhanced extraction API - ValueCollector added to make the collection of extracted results easier.

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/query/impl/getters/MultiResult.java [hunk 1]
```diff
@@ -74,6 +74,10 @@ public MultiResult() {
         this.results = new ArrayList<T>();
     }
 
+    public MultiResult(List<T> results) {
+        this.results = new ArrayList<T>(results);
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/query/impl/getters/MultiResult.java]: is it necessary to create a defensive copy? it's just internal API now.

File: hazelcast/src/main/java/com/hazelcast/query/impl/DefaultValueCollector.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+import com.hazelcast.query.extractor.ValueCollector;
+import com.hazelcast.query.impl.getters.MultiResult;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class DefaultValueCollector extends ValueCollector {
+
+    private List<Object> values = new ArrayList<Object>();
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/query/impl/DefaultValueCollector.java]: this seems to be quite wasteful. I can imagine a lot of extractor will produce just a single value.

File: hazelcast/src/main/java/com/hazelcast/query/impl/getters/MultiResult.java (hunk 1 above)
Comment (tombujok) [inline hazelcast/src/main/java/com/hazelcast/query/impl/getters/MultiResult.java]: It's was defensively copied with the assumption that the ValueCollector that calls this constructor will be recycled & reused so we didn't want to get this list reset when the Collector was recycled.
But for now, the collector is not reused so I can just take the reference as-is.

File: hazelcast/src/main/java/com/hazelcast/query/impl/DefaultValueCollector.java (hunk 2 above)
Comment (tombujok) [inline hazelcast/src/main/java/com/hazelcast/query/impl/DefaultValueCollector.java]: Good hint. Fixed.

File: hazelcast/src/main/java/com/hazelcast/query/impl/getters/MultiResult.java (hunk 1 above)
Comment (tombujok) [inline hazelcast/src/main/java/com/hazelcast/query/impl/getters/MultiResult.java]: Fixed

**Comment by jerrinot** [comment]
:+1:

**Comment by metanet** [comment]
:+1:

**Pull request merged**